### ISSUE TITLE
PLW0108 false positive if callable is defined later in the module

### ISSUE THREAD

**Issue by JelleZijlstra** [astral-sh/ruff#24704]
### Summary

On this code:

```python
x = {"a": lambda y: f(y)}

def f(y):
    pass
```

Ruff reports `Lambda may be unnecessary; consider inlining inner function (PLW0108)`.
But `f` is not defined yet at this point, so the code would fail with a NameError if we tried to remove the lambda.

### Version

0.15.10

**Comment by Lawson-Darrow** [comment]
Opened #25469. The lambda defers the `f` lookup to call time, so inlining `lambda y: f(y)` -> `f` raises NameError when `f` is bound later. The fix skips PLW0108 when a referenced name's binding starts after the lambda's end (a forward reference); comparing against the end keeps walrus-in-body cases like `lambda x: (string := str)(x)` reported.

**Comment by kimjune01** [comment]
PLW0108 should flag only lambdas that are actually safe to inline, but this isn't visible in its structure. Two snippets with identical syntax can have opposite correct answers. I checked both against CPython:

```python
# list comp builds eagerly, before `def f` -> inlining raises NameError. Must NOT flag.
cbs = [lambda x: f(x) for _ in range(1)]
def f(x): return x
cbs[0](1)

# genexpr body runs lazily, after `f` is bound -> inlining is safe. SHOULD flag.
cbs = (lambda x: f(x) for _ in range(1))
def f(x): return x
next(cbs)(1)
```

The flag depends on when its container evaluates.

Does the call target resolve to the same object when loaded eagerly at the lambda's position as it does lazily in the body? That is the safety condition, and it turns on execution order.  The binder is flow-insensitive (scope + source position), so position-based fixes like #25074 and #25469 each leave a different hole. The exact predicate is undecidable, so the boundary moves rather than closes.

I think this is the wrong thing to ask the linter's model. I suggest either:

- narrow the diagnostic to cases an immediate-load check proves safe, and document the rest;
- treat PLW0108 as a style signal about the forwarder shape, and drop the rewrite;
- defer the precise version to ty's flow-sensitive model, since definite assignment is what it computes.

Would a minimal patch help for whichever direction you pick?
